File: Graph/1971. Find if Path Exists in Graph/DFS.java
import java.util.*;

class Solution {
    public boolean validPath(int n, int[][] edges, int source, int destination) {
        HashMap<Integer, ArrayList<Integer>> link = new HashMap<>();
        for(int[] edge : edges){
            link.computeIfAbsent(edge[0], k -> new ArrayList<>()).add(edge[1]);
            link.computeIfAbsent(edge[1], k -> new ArrayList<>()).add(edge[0]);
        }

        boolean[] visited = new boolean[n];
        ArrayDeque<Integer> queue = new ArrayDeque<>();
        queue.addFirst(source);

        while(!queue.isEmpty()){
            int node = queue.poll();
            if(node == destination){return true;}
            if(!visited[node]){
                visited[node] = true;
                for(int neighbor : link.get(node)){
                    if(!visited[neighbor]){
                        queue.addLast(neighbor);
                    }
                }
            }
        }

        return false;
    }
}

File: 13-Graph/DFS/1971. Find if Path Exists in Graph/DFS.py
from collections import defaultdict, deque

class Solution:
    def validPath(self, n: int, edges: List[List[int]], source: int, destination: int) -> bool:
        link = defaultdict(list)
        for u, v in edges:
            link[u].append(v)
            link[v].append(u)
        visited = [False] * n
        queue = deque()
        queue.append(source)
        
        while queue:
            node = queue.popleft()
            if node == destination:
                return True
            
            if not visited[node]:
                visited[node] = True
                for neighbor in link[node]:
                    if not visited[neighbor]:
                        queue.append(neighbor)
        
        return False
